### PR TITLE
Add a check that tests if the sym file has grown above a percentage

### PR DESCRIPTION
**Pull request by mg-aquaq**
Add a check that tests whether todays symfile count is a certain percentage above the average sym file count from the previous n days.

### CODE REVIEW

File: code/dqc/symfilegrowth.q [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+/- Take the average symfile count from the past n days, then check that todays
+/- sym file count hasn't grown more than pct%.
+symfilegrowth: {[directory;ndays;pct]
+  /- Get handles to the DQE and the DQEDB. 
+  h:first exec w from .servers.SERVERS where proctype=`dqedb;
+  / Make sure we have enough days in the dqedb.
+  if[ndays>h"count .Q.pv";:(0b;"ERROR: number of days (ndays) exceeds number of available dates")];
+  / Get todays sym file count.
+  tc:dqeh"first exec resvalue from .dqe.resultstab where funct=`symcount";
```
Comment (ThomasSmyth) [inline code/dqc/symfilegrowth.q:11]: Is this used elsewhere (`dqeh`)? If the DQE restarts will this handle go stale? There are TorQ functions you could use here instead that should ensure you don't need to store a handle that could possibly go stale.

File: code/dqc/symfilegrowth.q [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  /- Get average sym file count from previous days, where select statement is
+  /- "select from resultstab where date>=.z.d-ndays,funct=`symfilecheck"
+  ac:exec avg resvalue from dqedbh(?;`resultstab;(((';~:;<);`date;(-;`.z.d;ndays));(=;`funct;enlist`symcount));0b;());
```
Comment (ThomasSmyth) [inline code/dqc/symfilegrowth.q:14]: Should be able to use a `select` statement here, looks like the only variable parameter is `ndays`.

Comment (mg-aquaq) [reply code/dqc/symfilegrowth.q:14]: So I could just do
```q
dqedbh "select from resultstab where date>=.z.d-",string[ndays],",funct=`symfilecheck"
```
?

Comment (ThomasSmyth) [reply code/dqc/symfilegrowth.q:14]: Looks good, easier to follow that way.

File: code/dqc/symfilegrowth.q (hunk 1 above)
Comment (mg-aquaq) [reply code/dqc/symfilegrowth.q:11]: I'll change it. Which torq functions?

**Comment by mg-aquaq** [comment]
Also realised I should add logging, so will do that now.

**Comment by mg-aquaq** [comment]
Should this function be looking at the previous n days, or the previous n business days? If so I will need to add in that functionality

File: code/dqc/symfilegrowth.q (hunk 1 above)
Comment (mg-aquaq) [reply code/dqc/symfilegrowth.q]: Used .servers.getservers, good suggestion from daniel

File: code/dqc/symfilegrowth.q [hunk 3]
```diff
@@ -0,0 +1,17 @@
+\d .dqc
+
+/- Take the average symfile count from the past n days, then check that todays
+/- sym file count hasn't grown more than pct%.
+symfilegrowth:{[ndays;pct]
+  .lg.o[`dqe;"Checking sym file has not grown more than ",string[pct],"%."];
+  /- Get handle to the DQEDB.
+  h:(exec first w from .servers.getservers[`proctype;`dqedb;()!();0b;1b]);
+  /- Make sure we have enough days in the dqedb.
+  if[ndays>h"count .Q.pv";:(0b;"ERROR: number of days (ndays) exceeds number of available dates")];
```
Comment (ThomasSmyth) [inline code/dqc/symfilegrowth.q:10]: Can give users more info here:
```suggestion
  if[ndays>c:h"count .Q.pv";:(0b;"ERROR: number of days (",string[ndays],") exceeds number of available dates (",string[c],") on disk")];
```

File: code/dqc/symfilegrowth.q [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  /- Get todays sym file count.
+  tc:first exec resvalue from .dqe.resultstab where funct=`symcount;
+  /- Get average sym file count from previous days.
+  ac:exec avg resvalue from h"select from resultstab where date>=.z.d-",string[ndays],",funct=`symcount";
+  /- Test whether the symfile growth is less than a pct, and return test status.
+  (b;"Sym file ",$[b:pct>100*(tc-ac)%ac;"has not";"has"]," grown more than ",string[pct],"% above the previous ",string[ndays],"-day average.")
```
Comment (ThomasSmyth) [inline code/dqc/symfilegrowth.q:16]: Small change here to log this message too:
```suggestion
  msg:"Sym file ",$[b:pct>100*(tc-ac)%ac;"has not";"has"]," grown more than ",string[pct],"% above the previous ",string[ndays],"-day average.";
  .lg.o[`dqe]msg;
  (b;msg)
```

File: code/processes/dqe.q [hunk 5]
```diff
@@ -79,10 +79,10 @@ writedownengine:{
 .dqe.currentpartition:.dqe.getpartition[];  /- initialize current partition
 
 
-.servers.CONNECTIONS:`tickerplant`rdb`hdb /- set to nothing so that is only connects to discovery
+.servers.CONNECTIONS:`tickerplant`rdb`hdb`dqedb /- set to nothing so that is only connects to discovery
```
Comment (ThomasSmyth) [inline code/processes/dqe.q:82]: This comment needs updated.

File: code/processes/dqe.q [hunk 6]
```diff
@@ -15,7 +15,7 @@ resultstab:([]procs:`$();funct:`$();table:`$();column:`$();resvalue:`long$());
 init:{
   .lg.o[`init;"searching for servers"];
   /- Open connection to discovery
-  .servers.startup[];
+  .servers.startupdependent[`dqedb;10];
```
Comment (ThomasSmyth) [inline code/processes/dqe.q:18]: Why this change?

Comment (mg-aquaq) [reply code/processes/dqe.q:18]: Due to this:

```
.servers.CONNECTIONS:`tickerplant`rdb`hdb`dqedb
```

in the previous comment. If it wasn't startupdependent, then it didn't start up properly.

File: code/dqc/symfilegrowth.q [hunk 7]
```diff
@@ -0,0 +1,19 @@
+\d .dqc
+
+/- Take the average symfile count from the past n days, then check that todays
+/- sym file count hasn't grown more than pct%.
+symfilegrowth:{[ndays;pct]
+  .lg.o[`dqe;"Checking sym file has not grown more than ",string[pct],"%."];
```
Comment (ThomasSmyth) [inline code/dqc/symfilegrowth.q:6]: Last nitpick, change is of this log message and the other on line 17 from `dqe` to something like `symfilegrowth`
